### DISCUSSION TITLE
Is there a way to create command "groups" for custom commands?

### DISCUSSION THREAD

**Discussion by TravisCarden** [ddev/ddev#6432] [Q&A]
I have several custom commands that I would like to set apart in `ddev help`. Is there a way to "namespace" them so that they appear in groups/sets/categories? An example of such functionality would be Drush, which looks like this:

```
$ drush
...
Available commands:
_global:
  archive:dump (ard)                                     Backup your code, files, and database into a single file.
  browse                                                 Display a link to a given path or open link in a browser.
  ...
cache:
  cache:clear (cc)                                       Clear a specific cache, or all Drupal caches.
  cache:get (cg)                                         Fetch a cached object and display it....
  ...
config:
  config:delete (cdel)                                   Delete a configuration key, or a whole object(s).
  config:edit (cedit)                                    Open a config file in a text editor. Edits are imported after closing editor.
...
```

```
$ drush --filter=pm
...Available commands:
pm:
  pm:install (in, install, en, pm-enable, pm:enable) Enable one or more modules.
  pm:list (pml)                                      Show a list of available extensions (modules and themes).
  pm:uninstall (un, pmu)                             Uninstall one or more modules and their dependent modules.
```

**Comment by sreichel** [comment]
I like that idea. 👍

**Reply by rfay** [reply]
It would be a good idea for DDEV to follow the lead of many CLIs that have this setup, but it hasn't happened and isn't currently planned.

**Comment by stasadev** [comment]
These groups can definitely be added, but it won't be as useful as in PHP, for example if you have some namespace like `cache` and you type `ddev cache`, you won't be prompted for commands from that namespace. So I decided not to create a PR.

Draft code, add this somewhere [here](https://github.com/ddev/ddev/blob/c2aca52a18687e678086dd232573cf51914dba56/cmd/ddev/cmd/commands.go#L278):

```go
// Add command group in "ddev help"
if val, ok := directives["Group"]; ok && val != "" {
    commandToAdd.GroupID = strings.TrimSuffix(val, ":") + ":"
    if !rootCmd.ContainsGroup(commandToAdd.GroupID) {
        rootCmd.AddGroup(&cobra.Group{ID: commandToAdd.GroupID, Title: commandToAdd.GroupID})
    }
}
```

Compile, and add this to some custom command:

```
## Group: My Commands
```

It will show in `ddev help` like this (I added it to `self-upgrade`):

```
$ ddev help    
Create and maintain a local web development environment.
Docs: https://ddev.readthedocs.io
Support: https://ddev.readthedocs.io/en/stable/users/support/

Usage:
  ddev [command]

My Commands:
  self-upgrade             Explain how to upgrade DDEV (global shell host container command)

Additional Commands:
  auth                     A collection of authentication commands
```

**Comment by sreichel** [comment]
In my opinion, it would be an improvement if the commands could be grouped visually.

**Comment by ChloeSartorelli** [comment]
This would be a good suggestion to make in the upstream dependency [`cobra`](https://github.com/spf13/cobra).
If it's implemented there, DDEV (and any other projects that use cobra) can take advantage of it.
But it would be difficult to implement in a meaningful way directly in DDEV itself.
